docs/error-codes.md

```diff
@@ -266,4 +472,154 @@ the error code. For example:
 
 #### `IL2049`: Unrecognized internal attribute 'attribute'
 
-- The internal attribute name 'attribute' being used in the xml is not supported by the linker, check the spelling and the supported internal attributes.
\ No newline at end of file
+- The internal attribute name 'attribute' being used in the xml is not supported by the linker, check the spelling and the supported internal attributes.
+
+#### `IL2050` Trim analysis: Making a generic type instantiation from 'type' which has 'DynamicallyAccessedMembersAttribute' on some of its generic parameters. ILLink currently doesn't analyze type values for generic parameters when making a generic type instantiation via 'System.Type.MakeGenericType'
+
+- Analysis of type parameters for `System.Type.MakeGenericType` is not yet implemented. If the open generic type `type` has `DynamicallyAccessedMembersAttribute` on any of its generic parameters, ILLink currently can't validate that the requirements are fulfilled by the calling method.  
+
+  ``` C#
+  class Lazy<[DynamicallyAccessedMembers(DynamicallyAccessedMemberType.PublicParameterlessConstructor)] T> 
+  {
+      // ...
+  }
+  
+  void TestMethod()
+  {
+      // IL2050 Trim analysis: Making a generic type instantiation from 'Lazy<T>' which has 
+      // 'DynamicallyAccessedMembersAttribute' on some of its generic parameters. ILLink currently doesn't analyze 
+      // type values for generic parameters when making a generic type instantiation via 'System.Type.MakeGenericType'
+      typeof(Lazy<>).MakeGenericType(new Type[] { typeof(TestType) });
+  }
+  ```
+
+  #### `IL2051` Trim analysis: Unrecognized type value used to called the method 'MakeGenericType'. It's not possible to guarantee the availability of requirements of such type.
+
+- If `System.Type.MakeGenericType` is called on an type which is not statically analyzable. If the actual type has generic parameters with `DynamicallyAccessedMembersAttribute` ILLink would be required to fulfill the requirements declared by those attributes, but since the ILLink doesn't know the type, it can't determine if such requirements exist.  
+
+  ``` C#
+  void TestMethod(Type type)
+  {
+      // IL2051 Trim analysis: Unrecognized value passed to the parameter 'typeArguments' of method 'System.Type.MakeGenericType'
+      type.MakeGenericType(new Type[] { typeof(TestType) });
+  }
+  ```
+
+#### `IL2052` Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(Type typeName)'. It's not possible to guarantee the availability of the target type.
```

Comment:
```suggestion
#### `IL2052` Argument type used for parameter 'typeName' of method 'System.Type.GetType(Type typeName)' cannot be determined.`
```

Reply:
I find the new message confusing - "Argument type" to me means the declared type of the argument, which in this case will pretty always be `System.Type` - and that one was determined (otherwise linker would not have a problem with it). It's the value passed to the argument which we could not figure out.
I guess there's also a question of terminology - argument or parameter?

Reply:
I think this is quite clear. From Wikipedia

> Unlike argument in usual mathematical usage, the argument in computer science is thus the actual input expression passed/supplied to a function, procedure, or routine in the invocation/call statement, whereas the parameter is the variable inside the implementation of the subroutine. 

I agree we need to settle on how we are going to identify the type and we are not consistent in other messages. In this case, it could be `Argument instance used for parameter ....`

Reply:
Thanks for the explanation. I will take a note and do this sweeping change in a separate PR (this is already getting out of hand). I will have another PR focusing just on the messages anyway.